### PR TITLE
Fix Kee Berry and Custap Berry + Tests

### LINKED ISSUE #3393: Battle claims that opponent gets the benefits of the player's Kee Berry
## Description
A Kee Berry raises the holder's Defense by one stage when hit by a physical move. This consumes the Kee Berry.

When the player has a Kee Berry attached to their Pokémon, and that Pokémon is damaged by a physical attack, the game claims that user of the physical attack gets the Defense boost from the Kee Berry, and then shows the same message.

(Checking the debug menu reveals that they do not actually get the Defense boost)

## Repo
<details>
<summary>data/scripts/debug.inc</summary>

```diff
Debug_Script_1::
+    givemon SPECIES_RAPIDASH_GALARIAN,100,ITEM_KEE_BERRY
    end

Debug_Script_2::
+    setwildbattle SPECIES_DONPHAN, 40, ITEM_NONE
+    dowildbattle
	end

Debug_Script_3::
+    setwildbattle SPECIES_HITMONCHAN, 20, ITEM_NONE
+    dowildbattle
	end
```

Run Script 1, then Script 2 or 3.

### Repro Steps
1. Rapidash is hit with physical attack by Donphan
2. Kee Berry activates
3. Message prints that Donphan gets Defense boost <-- bug
4. Message prints that Rapidash gets Defense boost
</details>

## Video
<details>
<summary>Video of bug</summary>

https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/ba833bc7-48aa-43d6-bef6-ee42efed0a44

</details>

### Version

1.6.1 (Default)

### Upcoming Version

70ed93e01b0f7910e2d96bc61be43bbbe8eaa433

### Discord contact info

pkmnsnfrn

### LINKED ISSUE #3394: Kee Berry raises Defense by 1 stage instead of 2
## Description
A Kee Berry raises the holder's Defense by one stage when hit by a physical move. This consumes the Kee Berry.

When the player has a Kee Berry attached to their Pokémon, and that Pokémon is damaged by a physical attack, the Pokémon holding the Kee Berry gets +2 Defense instead of +1. +2 Defense should only apply when the holding Pokémon's ability is Ripen.

## Repo
<details>
<summary>data/scripts/debug.inc</summary>

```diff
Debug_Script_1::
+    givemon SPECIES_RAPIDASH_GALARIAN,100,ITEM_KEE_BERRY
    end

Debug_Script_2::
+    setwildbattle SPECIES_DONPHAN, 40, ITEM_NONE
+    dowildbattle
	end

Debug_Script_3::
+    setwildbattle SPECIES_HITMONCHAN, 20, ITEM_NONE
+    dowildbattle
	end
```

Run Script 1, then Script 2 or 3.

### Repro Steps
1. Rapidash is hit with physical attack by Hitmonchan
2. Kee Berry activates
3. Message prints that Rapidash gets Defense boost
4. Check Debug menu - Rapidash has +2 DEF <- bug
</details>

## Video
<details>
<summary>Video of bug</summary>

https://github.com/rh-hideout/pokeemerald-expansion/assets/77138753/e1aa4b3e-651f-4cf8-aaaa-971a26abaee2

</details>

### Version

1.6.1 (Default)

### Upcoming Version

70ed93e01b0f7910e2d96bc61be43bbbe8eaa433

### Discord contact info

pkmnsnfrn

### PR DESCRIPTION
**Pull request by AlexOn1ine**
Fixes  #3394
Fixes  #3393

Also fixes Custap Berry

### CODE REVIEW

File: test/battle/hold_effect/attack_up.c [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        MESSAGE("Using Liechi Berry, the Attack of Bellsprout rose!");
+    } THEN {
+        EXPECT_EQ(player->statStages[STAT_ATK], DEFAULT_STAT_STAGE + 1);
+    }
+}
+
+SINGLE_BATTLE_TEST("Liechi Berry raises Attack by one stage when HP drops to 1/4 or below if holder has Ripen")
+{
+    GIVEN {
+        PLAYER(SPECIES_APPLIN) { MaxHP(160); HP(80); Ability(ABILITY_RIPEN); Item(ITEM_LIECHI_BERRY); }
```
Comment (DizzyEggg) [inline test/battle/hold_effect/attack_up.c:59]: Assume gen8 mons

File: test/battle/hold_effect/critical_hit_up.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_LANSAT_BERRY].holdEffect == HOLD_EFFECT_CRITICAL_UP;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/critical_hit_up.c:6]: Missing ASSUME();

File: test/battle/hold_effect/critical_hit_up.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    gBattleMoves[MOVE_DRAGON_RAGE].effect == EFFECT_DRAGON_RAGE;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/critical_hit_up.c:7]: Missing ASSUME();

File: test/battle/hold_effect/critical_hit_up.c [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+SINGLE_BATTLE_TEST("Lansat Berry raises the holder's critical-hit-ratio by two stage when HP drops to 1/4 or below")
```
Comment (DizzyEggg) [inline test/battle/hold_effect/critical_hit_up.c:10]: by two stage**s**

File: test/battle/hold_effect/critical_hit_up.c [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        else
+        {
+            ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, player);
+            MESSAGE("Wobbuffet used Lansat Berry to get pumped!");
+        }
+    }
+}
+
+SINGLE_BATTLE_TEST("Lansat Berry raises the holder's critical-hit-ratio by two stage when HP drops to 1/2 or below")
```
Comment (DizzyEggg) [inline test/battle/hold_effect/critical_hit_up.c:37]: by two stages

File: test/battle/hold_effect/critical_hit_up.c [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_DRAGON_RAGE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_DRAGON_RAGE, opponent);
+        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, player);
+        MESSAGE("Bellsprout used Lansat Berry to get pumped!");
+    }
+}
+
+SINGLE_BATTLE_TEST("Lansat Berry raises the holder's critical-hit-ratio by two stage when HP drops to 1/4 or below")
```
Comment (DizzyEggg) [inline test/battle/hold_effect/critical_hit_up.c:51]: by two stages

File: test/battle/hold_effect/custap_berry.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_CUSTAP_BERRY].holdEffect == HOLD_EFFECT_CUSTAP_BERRY;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/custap_berry.c:6]: Missing ASSUME

File: test/battle/hold_effect/custap_berry.c [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+SINGLE_BATTLE_TEST("Custap Berry")
```
Comment (DizzyEggg) [inline test/battle/hold_effect/custap_berry.c:9]: maybe a better test name?

File: test/battle/hold_effect/custap_berry.c [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        TURN { MOVE(player, MOVE_TACKLE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, player);
+        MESSAGE("Wobbuffet can act faster, thanks to Custap Berry!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, player);
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_CELEBRATE, opponent);
+    }
+}
+
+SINGLE_BATTLE_TEST("Custap Berry 2")
```
Comment (DizzyEggg) [inline test/battle/hold_effect/custap_berry.c:24]: maybe a better test name?

File: test/battle/hold_effect/defense_up.c [hunk 10]
```diff
@@ -0,0 +1,70 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_GANLON_BERRY].holdEffect == HOLD_EFFECT_DEFENSE_UP;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/defense_up.c:6]: missing ASSUME

File: test/battle/hold_effect/defense_up.c [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    gBattleMoves[MOVE_DRAGON_RAGE].effect == EFFECT_DRAGON_RAGE;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/defense_up.c:7]: missing ASSUME

File: test/battle/hold_effect/jaboca_berry.c [hunk 12]
```diff
@@ -0,0 +1,44 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_JABOCA_BERRY].holdEffect == HOLD_EFFECT_JABOCA_BERRY;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/jaboca_berry.c:6]: missing ASSUME

File: test/battle/hold_effect/kee_berry.c [hunk 13]
```diff
@@ -0,0 +1,44 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_JABOCA_BERRY].holdEffect == HOLD_EFFECT_JABOCA_BERRY;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/kee_berry.c:6]: missing ASSUME

File: test/battle/hold_effect/maranga_berry.c [hunk 14]
```diff
@@ -0,0 +1,43 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_MARANGA_BERRY].holdEffect == HOLD_EFFECT_MARANGA_BERRY;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/maranga_berry.c:6]: missing ASSUME

File: test/battle/hold_effect/micle_berry.c [hunk 15]
```diff
@@ -0,0 +1,68 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_MICLE_BERRY].holdEffect == HOLD_EFFECT_MICLE_BERRY;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/micle_berry.c:6]: missing ASSUME

File: test/battle/hold_effect/micle_berry.c [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    gBattleMoves[MOVE_DRAGON_RAGE].effect == EFFECT_DRAGON_RAGE;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/micle_berry.c:7]: missing ASSUME

File: test/battle/hold_effect/random_stat_up.c [hunk 17]
```diff
@@ -0,0 +1,11 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_STARF_BERRY].holdEffect == HOLD_EFFECT_RANDOM_STAT_UP;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/random_stat_up.c:6]: missing ASSUME

File: test/battle/hold_effect/rowap_berry.c [hunk 18]
```diff
@@ -0,0 +1,44 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_KEE_BERRY].holdEffect == HOLD_EFFECT_KEE_BERRY;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/rowap_berry.c:6]: missing ASSUME

File: test/battle/hold_effect/special_attack_up.c [hunk 19]
```diff
@@ -0,0 +1,70 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_PETAYA_BERRY].holdEffect == HOLD_EFFECT_SP_ATTACK_UP;
+    gBattleMoves[MOVE_DRAGON_RAGE].effect == EFFECT_DRAGON_RAGE;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/special_attack_up.c:7]: missing ASSUME

File: test/battle/hold_effect/special_defense_up.c [hunk 20]
```diff
@@ -0,0 +1,70 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_APICOT_BERRY].holdEffect == HOLD_EFFECT_SP_DEFENSE_UP;
+    gBattleMoves[MOVE_DRAGON_RAGE].effect == EFFECT_DRAGON_RAGE;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/special_defense_up.c:7]: missing ASSUME

File: test/battle/hold_effect/speed_up.c [hunk 21]
```diff
@@ -0,0 +1,70 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    gItems[ITEM_SALAC_BERRY].holdEffect == HOLD_EFFECT_SPEED_UP;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/speed_up.c:6]: missing ASSUME

File: test/battle/hold_effect/speed_up.c [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    gBattleMoves[MOVE_DRAGON_RAGE].effect == EFFECT_DRAGON_RAGE;
```
Comment (DizzyEggg) [inline test/battle/hold_effect/speed_up.c:7]: missing ASSUME

**Review by DizzyEggg — CHANGES_REQUESTED** [review]

**Comment by AlexOn1ine** [comment]
Changes applied.

File: test/battle/hold_effect/kee_berry.c [hunk 23]
```diff
@@ -0,0 +1,44 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gItems[ITEM_JABOCA_BERRY].holdEffect == HOLD_EFFECT_JABOCA_BERRY);
```
Comment (DizzyEggg) [inline test/battle/hold_effect/kee_berry.c:6]: This whole file is for jaboca berry, but should be for kee berry.

File: test/battle/hold_effect/rowap_berry.c [hunk 24]
```diff
@@ -0,0 +1,44 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gItems[ITEM_KEE_BERRY].holdEffect == HOLD_EFFECT_KEE_BERRY);
```
Comment (DizzyEggg) [inline test/battle/hold_effect/rowap_berry.c:6]: This file is supposed to be for rowap berry, not kee berry.

**Review by DizzyEggg — CHANGES_REQUESTED** [review]

File: test/battle/hold_effect/kee_berry.c (hunk 23 above)
Comment (AlexOn1ine) [reply test/battle/hold_effect/kee_berry.c]: Should be fixed.

File: test/battle/hold_effect/rowap_berry.c (hunk 24 above)
Comment (AlexOn1ine) [reply test/battle/hold_effect/rowap_berry.c]: Should be fixed.

**Comment by AlexOn1ine** [comment]
There is a new line missing in the `maranga_berry.c` file. I'll fix it right away if everything else is fine.

File: test/battle/hold_effect/kee_berry.c [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_KEE_BERRY); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_SWIFT); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_SWIFT, player);
+        HP_BAR(opponent);
+        NOT ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, opponent);
+        NOT MESSAGE("Using Kee Berry, the Defense of Foe Wobbuffet rose!");
+    }
+}
```
Comment (DizzyEggg) [inline test/battle/hold_effect/kee_berry.c:60]: newline

File: test/battle/hold_effect/maranga_berry.c [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+        OPPONENT(SPECIES_WOBBUFFET) { Item(ITEM_MARANGA_BERRY); }
+    } WHEN {
+        TURN { MOVE(player, MOVE_TACKLE); }
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_TACKLE, player);
+        HP_BAR(opponent);
+        NOT ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_HELD_ITEM_EFFECT, opponent);
+        NOT MESSAGE("Using Maranga Berry, the Sp. Def of Foe Wobbuffet rose!");
+    }
+}
```
Comment (DizzyEggg) [inline test/battle/hold_effect/maranga_berry.c:60]: newline

**Review by DizzyEggg — CHANGES_REQUESTED** [review]
Other than it's good

File: test/battle/hold_effect/maranga_berry.c (hunk 26 above)
Comment (AlexOn1ine) [reply test/battle/hold_effect/maranga_berry.c]: fixed

File: test/battle/hold_effect/kee_berry.c (hunk 25 above)
Comment (AlexOn1ine) [reply test/battle/hold_effect/kee_berry.c]: fixed

**Review by DizzyEggg — APPROVED** [review]

**Pull request merged**